[#26198] Modify the bot_avatar_row.hbs template to Include the download button in the keyboard navigation.

EdgarRamirezFuentes opened the pull request:
<!-- Describe your pull request here.-->
This PR modifies the bot_avatar_row.hbs template to include the download button in the keyboard navigation.
I added the tabindex attribute in the anchor tag to focust the button in the keyboard navigation.

Fixes: <!-- Issue link, or clear description.-->
https://github.com/zulip/zulip/issues/26198

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

<img width="384" alt="Captura de pantalla 2023-07-06 a la(s) 21 36 19" src="https://github.com/zulip/zulip/assets/50359137/45e74a4a-3dda-4dfb-9a20-9331113957f9">

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

EdgarRamirezFuentes commented:
@sbansal1999 does this fix your issue?

amanagr commented:
LGTM

sbansal1999 commented:
@EdgarRamirezFuentes Congratulations on your first contribution to Zulip. :tada: 

The change looks good to me. The border when the `Download zuliprc` button is focused is different from other buttons. So, if you can fix that it would be great.

<details>

<summary> Screencast of the current behaviour</summary>

[Screen-recording](https://github.com/zulip/zulip/assets/35286603/108b0521-6db3-421f-8c7c-eddbad67d57f)

</details>

Also, I would recommend rewriting the commit message to align with our [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html#commit-discipline). Specifically:

- The commit message could be something like `bots_settings: Make "Download zuliprc" button keyboard focusable`. 
- The body of the commit message should explain why and how the change was made and should be line-wrapped to about 68 characters.
- Just like you have mentioned `Fixes #26198` in the pull request, make sure to add the same to the end of the commit message. This will result in that issue getting closed automatically when the code is merged.

EdgarRamirezFuentes commented:
Thanks for the feedback!
I'll be working on the border issue, and rewriting the commit to align the commit discipline.

EdgarRamirezFuentes commented:
@sbansal1999 I hope you are doing well!
I modified the commit message and the commit message body as you mentioned. Is it correct?
And... I also fixed the style issue in the "Download zuliprc" button.

EdgarRamirezFuentes commented:
I also realized that some anchor tags use the attribute [type="submit"] and that is not correct.
Should I delete this attribute to all the anchor tags using it? 
I didn’t delete any attribute because it could lead to some bug, but I wanted to know if it's ok if those attributes are deleted.

I checked the anchor tag attributes here:
https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a

timabbott commented on web/templates/settings/bot_avatar_row.hbs:
```diff
@@ -8,8 +8,10 @@
                 </button>
-                <a type="submit" download="{{zuliprc}}" class="btn download_bot_zuliprc tippy-zulip-delayed-tooltip" data-tippy-content="{{t 'Download zuliprc' }}" data-email="{{email}}">
-                    <i class="fa fa-download sea-green" aria-hidden="true"></i>
+                <a type="submit" download="{{zuliprc}}"  class="btn download_bot_zuliprc tippy-zulip-delayed-tooltip" data-tippy-content="{{t 'Download zuliprc' }}" data-email="{{email}}">
```
The extra space before "class" should be removed.

sbansal1999 replied:
Since we are adding a button tag inside the anchor tag there is no need to have this `btn` class.

This also leads to a scroll bar being added to the box as well as irregular spacing between the icons.

<details>

<summary> See scrollbar and irregular spacing</summary>

![image](https://github.com/zulip/zulip/assets/35286603/d7dec6b5-8c49-4bef-87c8-8ee1101266d9)

</details>

sbansal1999 commented on web/templates/settings/bot_avatar_row.hbs:
```diff
@@ -8,8 +8,10 @@
-                <a type="submit" download="{{zuliprc}}" class="btn download_bot_zuliprc tippy-zulip-delayed-tooltip" data-tippy-content="{{t 'Download zuliprc' }}" data-email="{{email}}">
-                    <i class="fa fa-download sea-green" aria-hidden="true"></i>
+                <a type="submit" download="{{zuliprc}}"  class="btn download_bot_zuliprc tippy-zulip-delayed-tooltip" data-tippy-content="{{t 'Download zuliprc' }}" data-email="{{email}}">
+                    <button type="submit" class="btn tippy-zulip-delayed-tooltip">
```
Adding the `tippy-zulip-delayed-tooltip` class here leads to an empty tooltip being created. So we should remove it.

<details>

<summary> An empty tooltip</summary>

![image](https://github.com/zulip/zulip/assets/35286603/5ffb59dd-e1be-448f-a85c-0fdb89e0969d)

</details>

sbansal1999 reviewed (commented):
@EdgarRamirezFuentes This is some good work. 

I have left a few comments.  Feel free to tag me when you have addressed them or if you have any doubts.

EdgarRamirezFuentes commented:
Hi, @sbansal1999! 
Thanks for your feedback! 
I made the changes you mentioned before, but I realized that the text-decoration in the anchor tag shows a kind of blue underscore when you hover it. 
<details>
<summary>Anchor tag with text decoration</summary>
<img width="256" alt="Captura de pantalla 2023-07-14 a la(s) 20 58 39" src="https://github.com/zulip/zulip/assets/50359137/b40a48df-674a-4bf9-969c-b8f43de11d85">
</details>

So, I decided to create the class `.no-text-decoration` in `/web/styles/settings.css`, which helps to delete this "underscore" by setting the text-decoration as none.
<details>
<summary>no-text-decoration class</summary>
<img width="536" alt="Captura de pantalla 2023-07-14 a la(s) 21 19 35" src="https://github.com/zulip/zulip/assets/50359137/df95dc10-40e4-47e0-a34b-6249bb32ac47">
</details>


<details>
<summary>Anchor tag with no text decoration</summary>
<img width="355" alt="Captura de pantalla 2023-07-14 a la(s) 21 16 10" src="https://github.com/zulip/zulip/assets/50359137/fc0fcace-3661-4e86-a9fe-c6a6296b9c21">
</details>

Is this a correct approach?

I'll wait for your feedback to commit and push this changes.

Have a nice day!

sbansal1999 commented:
> So, I decided to create the class .no-text-decoration in /web/styles/settings.css, which helps to delete this "underscore" by setting the text-decoration as none.

@EdgarRamirezFuentes  Small utility classes like these are pretty common and are used in various places. A quick grep can tell you if a class that does this is already present or not.

So instead of creating `.no-text-decoration ` class, you can use the `no-underline` class defined in [components.css](https://github.com/zulip/zulip/blob/main/web/styles/components.css) and that should do the same thing.

https://github.com/zulip/zulip/blob/b29ec4d62ef38a224467659c4f729ac9909ab69b/web/styles/components.css#L58-L61

EdgarRamirezFuentes commented:
@sbansal1999 Thanks for your help! The class no-underline worked.
I'll keep in mind your advice for my next contributions.

<details>
<summary>Hover on "Download zuliprc button"</summary>
<img width="230" alt="Captura de pantalla 2023-07-15 a la(s) 8 18 43" src="https://github.com/zulip/zulip/assets/50359137/4f6e46c2-9cb3-43c7-8d68-92e0c055f7e7">
</details>

<details>
<summary>Focus on "Download zuliprc button"</summary>
<img width="363" alt="Captura de pantalla 2023-07-15 a la(s) 8 23 37" src="https://github.com/zulip/zulip/assets/50359137/5bed0aa2-98af-4b01-86f2-c28be6301aaa">
</details>

sbansal1999 commented on web/templates/settings/bot_avatar_row.hbs:
```diff
@@ -8,8 +8,10 @@
                 </button>
-                <a type="submit" download="{{zuliprc}}" class="btn download_bot_zuliprc tippy-zulip-delayed-tooltip" data-tippy-content="{{t 'Download zuliprc' }}" data-email="{{email}}">
-                    <i class="fa fa-download sea-green" aria-hidden="true"></i>
+                <a tabindex="-1" download="{{zuliprc}}" class="download_bot_zuliprc tippy-zulip-delayed-tooltip no-underline" data-tippy-content="{{t 'Download zuliprc' }}" data-email="{{email}}">
```
You can remove the `tabindex` attribute from here. It's not needed anymore as the button tag is there.

EdgarRamirezFuentes replied:
@sbansal1999 I added that tabindex attribute, because I realized there are moments that the anchor tag is focused in the koeyboard navigation (not always) and it breaks the styles of the button. So, adding the negative tabindex the anchor tab is disabled in the keyboard navigation.

Should I delete that tabindex anyways?

Thanks for the info! I'll squash the commits into a single one.

sbansal1999 replied:
> I realized there are moments that the anchor tag is focused in the koeyboard navigation (not always) and it breaks the styles of the button

I tried removing the attribute and it worked for me. Can you describe the exact steps to reproduce this?

EdgarRamirezFuentes replied:
I tried to reproduce it, but It does not focus the anchor tag anymore. Maybe It happened with some changes I did before adding the button tag. Anyways, this last commit Fixes #26198.  

I also squashed the 4 commits into 1. The Git guide was useful and I learnt something new. Thanks!

Have a nice day! :D

sbansal1999 commented:
@EdgarRamirezFuentes Left one more small comment. 

Also, I saw there are right now three commits in the PR. I think you should [squash](https://zulip.readthedocs.io/en/latest/git/fixing-commits.html#squashing-commits) the commit into a single one. Our [Git guide](https://zulip.readthedocs.io/en/latest/git/index.html#git-guide) is a useful resource for getting information on all these things.

EdgarRamirezFuentes commented:
@sbansal1999 I realized that the commit failed in the Ubuntu 20.04. Did I do something wrong?

sbansal1999 commented:
@EdgarRamirezFuentes The failed test is actually a puppeteer test which can be flaky. So, I would recommend running the test locally by following the instructions at the end of the log and see if it fails locally as well; which I think it won't as the changes you have made are not related to the failed test.

After running it locally, you can force push by amending the recent commit and then the CI will start again, and hopefully passes.

EdgarRamirezFuentes commented:
Hi, @sbansal1999!
I ran the tests locally, and as expected it passed all the tests.
I forced pushed the commit, I hope it passes the tests.

I'll be pending of the result.

Have a nice day!

sbansal1999 commented on web/templates/settings/bot_avatar_row.hbs:
```diff
@@ -8,8 +8,10 @@
                 </button>
-                <a type="submit" download="{{zuliprc}}" class="btn download_bot_zuliprc tippy-zulip-delayed-tooltip" data-tippy-content="{{t 'Download zuliprc' }}" data-email="{{email}}">
-                    <i class="fa fa-download sea-green" aria-hidden="true"></i>
+                <a download="{{zuliprc}}" class="download_bot_zuliprc tippy-zulip-delayed-tooltip no-underline" data-tippy-content="{{t 'Download zuliprc' }}" data-email="{{email}}">
```
I don't think there is a need to remove `type="submit"` from here.

sbansal1999 reviewed (commented):
@EdgarRamirezFuentes  I missed this one earlier. Just one small change and then this should be good to go.

EdgarRamirezFuentes commented:
@sbansal1999 Done!
Thanks for your feedback!

Have a nice day!

sbansal1999 reviewed (approved):
LGTM. Marking this for integration review.

nit: Commit message most probably should be:

```
bots_settings: Make "Download zuliprc" button keyboard focusable.

Added a button tag to make the "Download zuliprc" button 
keyboard focusable.

Fixes #26198.
```

sbansal1999 commented:
@zulipbot add "integration review"
